### DISCUSSION TITLE
Questions about the galaxy ellipticity in the CosmoDC2 mock catalog

### DISCUSSION THREAD

**Discussion by yuemh** [LSSTDESC/desc-data-portal#82] [Q&A] [answered]
Hi,

This is Minghao Yue who is a grad student at the University of Arizona. I am working with Prof. Xiaohui Fan on modeling the population of gravitationally lensed quasars at high redshift. We are using the CosmoDC2 mock catalog to describe the deflector galaxy population, and we encountered some problems with the mock galaxy ellipticities.

Briefly speaking, we feel that the ellipticities of the mock galaxies are too low to be real. The majority of the mock galaxies are concentrated at e<0.1. To make sure that we are doing things correctly, we use the "ellipticity_true" column. We also checked the disk and bulge ellipticities, and the issue is similar. For example, this is what I get for a sky area of ~2 deg^2:

![image](https://user-images.githubusercontent.com/28637952/130310419-0c623d54-f0b8-470b-a2f1-7b2b26502357.png)

I have attached a document that contains detailed descriptions of my questions and some tests I made. I am wondering if anyone would like to provide comments and/or suggestions about this. I would be very grateful if you would like to help!

Best,
Minghao
[questions.pdf](https://github.com/LSSTDESC/desc-data-portal/files/7024851/questions.pdf)

**Comment by yymao** [comment]
Welcome Minghao!  I'm tagging @aphearin as he may have an answer to your question. 

In the meantime, I wonder if you apply any cuts on the galaxy magnitude when making this plot?

**Reply by yuemh** [reply]
Thanks a lot for your reply! For the figures I show here, I don't apply any cuts (except for RA and DEC). I have been working at the massive end, though, since I use them as deflector galaxies in strong lensing systems. The case is the same for massive (and thus bright) galaxies.

I notice that there is a Github issue regarding ellipticities:
https://github.com/LSSTDESC/cosmodc2/issues/10
And magnitude cuts are discussed there. The figures under that issue also show a huge concentration at e~0 (note the color map is in log scale).

For your convenience, I also paste Figure 7 from Joamichi+2013 here, which is the reference for the ellipticity distribution in the CosmoDC2 paper. It doesn't look like the sharp peak at e~0 is a result of magnitude cuts, since the two distributions differ so significantly.

![image](https://user-images.githubusercontent.com/28637952/130331594-95e02da2-0cb9-4c45-9e50-1de8d1ed0903.png)

Again, thank you so much for your help!

**Comment by evevkovacs** [comment, marked as answer]
Hi Minhao, Here is the ellipticity distribution from the cosmoDC2 validation paper which should be on the archive soon. 
![image](https://user-images.githubusercontent.com/7797837/130339694-7dad0ce0-acf3-4905-ba1d-813719e3a7f7.png)
As you can see, cosmoDC2, (with magnitude cuts) agrees well (as it is supposed to because the model for the cosmoDC2 ellipticities were tuned on the COSMOS data. The selection cuts that were made to obtain these plots   attempt to mimic those in the data as far as possible. In addition to the magnitude cuts, there is a B/T cut to select morphology. The other thing to watch out for is the definition of ellipticity that is used.

**Reply by yuemh** [reply]
Hi Eve,

Thank you so much! It helps a lot. Then I guess something's wrong with the way I get the ellipticity. May I ask for your help to check what might be wrong?

I am using the CosmoDC2 catalog v1.1.4 from here: https://portal.nersc.gov/project/lsst/cosmoDC2/, then access the data using GCRCatalog like this:

#### code start ####

import numpy as np
import matplotlib.pyplot as plt

from astropy.cosmology import FlatLambdaCDM
from astropy import units as u
from astropy import constants as c
from astropy.io import fits
from astropy.table import Table, vstack, hstack
import re

defaultcosmo = FlatLambdaCDM(H0=71, Om0=0.2648)

import GCRCatalogs

config_overwrite = dict(
    catalog_root_dir='/xdisk/xfan/yuemh/CosmoDC2/rawcat/portal.nersc.gov/project/lsst/cosmoDC2/',
    check_file_list_complete=True,
    check_size=True,
)

gc = GCRCatalogs.load_catalog('cosmoDC2_v1.1.4_image', config_overwrite)
quantities = ['ellipticity_true', 'redshift', 'galaxy_id', 'size_true', 'size_minor_true', \
                 'ellipticity_disk_true',\
                 'ellipticity_bulge_true', 'size_disk_true', 'size_minor_disk_true',\
                 'size_bulge_true', 'size_minor_bulge_true']

filters = ['ra>60', 'ra<62', 'dec<-35', 'dec>-37']
result = gc.get_quantities(quantities, filters=filters)
tbl = Table(result)
tbl.write('./elliptesttbl.fits', overwrite=True)

#### code end ####

I use the column "ellipticity_true" as the ellipticity. I also tried q = size_minor_true/size_true and e = 1-q / 1+q, which gives me the same result.

May I ask what is the correct way to get the data? Anything wrong with what I did?

Again, many many thanks for your help!

**Reply by yuemh** [reply]
Hi Eve,

I also tested gathering data from TAP and apply a luminosity cut. The result is like this:
![taptest](https://user-images.githubusercontent.com/28637952/130366158-a55e9162-14d7-4317-87b3-5275e5d32b7c.png)

It still looks weird.

Since there is no V-band absolute magnitude in the published CosmoDC2 catalog, I use LSST r instead. I still see a sharp peak at e~0.

In addition, according to the CosmoDC2 paper, the disk ellipticities are drawn from a Johnson SB distribution, the parameters of which do not evolve with magnitudes. So I am expecting that the disk ellipticities (the "ellipticity_disk_true" column") should look normal regardless of the subset I am checking -- but it is not.

May I ask if how do you get the ellipticities shown in the figure you sent to me? Just want to check what might be wrong with my data acquisition.

Thank you so much!

**Reply by evevkovacs** [reply]
Are you able to look at [the code for the DESCQA test](https://github.com/LSSTDESC/descqa/blob/master/descqa/EllipticityDistribution.py), because that would be easiest way to answer your questions.   A few quick notes: there is a V-band luminosity in the catalog native quantities, we also use a B/T cut and ellipticity is defined as sqrt((1-q^2)/(1+q^2)) with q=b/a and uses size and size_minor for a and b respectively:
```
'e_squared':{'possible_quantities':[['size', 'size_true'], ['size_minor', 'size_minor_true']],
                                                         'function':self.e_squared,
                                                         'xaxis_label': r'$e = \sqrt{(1-q^2)/(1+q^2)}$',
```
So I think that the main issue is that you need to use the other definition of ellipticity (matches the one used in the COSMOS data)

**Reply by yuemh** [reply]
Hi Eve,

This is super useful! Thanks a lot!

I'll check the code carefully. At this time, though, I realized one key problem: I might have the wrong version of the catalog, because my version does not have columns "size", "size_minor", and the V-band mag. May I ask where can I get the most recent version of CosmoDC2?

FYI, here is where I got it: https://portal.nersc.gov/project/lsst/cosmoDC2/

Thanks,
Minghao

**Comment by evevkovacs** [comment]
Hi Minghao, No you don't have the wrong version. I assume that you are using the GCR catalog reader. The quantities listed by default are not all the quantities available. You need to use the list_all_native_quantities() to see everything in the catalog. The column name you want is 
`'otherLuminosities/totalLuminositiesStellar:V:rest:dustAtlas'`. To get an AB magnitude just calculate -2.5*log10 of this.

**Reply by yuemh** [reply]
Hi Eve,

I tried running the validation code on my machine. What I did was cloning the DESCQA repository and adding the following lines at the bottom of the EllipticityDistribution.py file:

## Code Start

    import GCRCatalogs
    def main():
        percentage = np.arange(0, 101, 1)
        ed = EllipticityDistribution(validation_percentile_points=percentage, validation_percentile_ranges=[[0,1]])
        config_overwrite = dict(
                catalog_root_dir='/xdisk/xfan/yuemh/CosmoDC2/rawcat/portal.nersc.gov/project/lsst/cosmoDC2/',
                check_file_list_complete=True,
                check_size=True)
        gc = GCRCatalogs.load_catalog('cosmoDC2_v1.1.4_image', config_overwrite)
        ed.run_on_single_catalog(gc,'CosmoDC2','../../../')
    main()

## Code End 

I don't change anything else. Here is the output Nvse figure:

![Nvse_m_i_M_r](https://user-images.githubusercontent.com/28637952/130531521-66403041-d04b-4f27-ba01-9d41ce2320bf.png)

This one also has a sharp peak at e~0 and does not look similar to the Figure you shared with me.

I want to make sure that I am running the code in the correct way. But anyway, if I understand the code correctly, the plotted quantity is simply the "ellipticity_true" column. I don't think it is physical to have it peaked at e~0 no matter what the luminosity range is.

Do you have any ideas about what could be wrong? and may I ask if it would be possible for me to have a discussion via zoom? That will be extremely helpful.

Thank you so much again for your help!

**Reply by yuemh** [reply]
OK, I figured out how to include the observations. Now it looks closer to what you sent to me. There is a normalization issue, but I don't think it matters.
![Nvse2_m_i_M_V](https://user-images.githubusercontent.com/28637952/130543252-5d0e5b06-bc6a-4f23-b6fb-6480b78452b0.png)

**Reply by yuemh** [reply]
I am still confused, though. I will stop worrying about the comparison with observations at this moment. But I see this description from Korytov et al. (2019) (the main CosmoDC2 paper):

![image](https://user-images.githubusercontent.com/28637952/130543482-d68083a1-c5ea-46f6-bf8c-b47481d6e800.png)

If I get it correctly, the distribution of ellipticity is the same for all disk components and evolves with the luminosity for bulge components.

I plotted the Johnson SB distribution functions, for the disk parameters and the bulge parameters with different magnitudes, and the  result looks like this:

![analytical](https://user-images.githubusercontent.com/28637952/130543791-269ca829-33df-44a9-ad37-0bafbbb1b7de.png)


It looks good, but it differs significantly from the distribution of "ellipticity_true" in the mock catalog.

I don't think magnitude cuts and B/T cuts matters here because (1) the distribution of disk ellipticity should not be influenced by these factors, according to the paper. (2) None of the Johnson SB functions have a peak at e=0, so I don't expect any combination or subsets can show a very significant concentration at e=0 (as shown in the first figure I uploaded).

I would be very grateful if you would like to provide any suggestions!

**Reply by yuemh** [reply]
Hi Eve,

Thank you so much again for all your advice! The QA code you shared with me is especially useful. Now I think I have figured out what's going on; it's the definition of ellipticity. If I use e=sqrt[(1-q^2)/(1+q^2)] I get the expected result.

I still wish to double-check a few things, though. I don't find the data from Joamichi et al. (2013), so I guess that you get it via private communication. The definition of ellipticity in Joamichi+2013 is (1-q^2)/(1+q^2) if I get it correctly. So when I use e=sqrt[(1-q^2)/(1+q^2)] and get a similar distribution as Figure 7 in Joamichi+2013, I am a bit concerned about it.

Actually, if we use axis ratios (q = size_minor_true/size_true) as the indicator, the sharp peak at q=1 still looks somewhat unreal. 

![artest](https://user-images.githubusercontent.com/28637952/130585092-cd1a5417-3eb6-47ea-9e3d-743a466d20d2.png)

I am wondering if the distribution of sqrt[(1-q^2)/(1+q^2)] should actually be (1-q^2)/(1+q^2), i.e., I am guessing

sqrt[(1-q_now^2)/(1+q_now^2)] = (1-q_proposed^2)/(1+q_proposed^2)

If I change q_now (the current values) to q_proposed according to the equation above, I get this distribution:

![arproposed](https://user-images.githubusercontent.com/28637952/130585400-14561358-ffc6-4fbe-8c7e-ab3c9ab2165e.png)

which looks much better.

Anyway, now I think I know what causes the difference that makes me confused. Just wondering if you would like to provide any comments about the definition of ellipticity, i.e., why sqrt[(1-q^2)/(1+q^2)] is used in the QA, and which quantity is the Johnson SB distribution in the paper describing. If you would like to give it a double check, I would be very grateful.

Thank you again for all your help!

**Reply by evevkovacs** [reply]
It will take me a bit of time to follow up on the issues with the e definition due to other time commitments. I can give you some immediate information that should help though.  The total ellipticity is a luminosity weighted sum of the disk and bulge ellipticities (each of which come from the model) and the luminosities that are used are the SDSS ones:
```
 lum_disk = hgroup['SDSS_filters/diskLuminositiesStellar:SDSS_r:rest'].value
    lum_sphere = hgroup['SDSS_filters/spheroidLuminositiesStellar:SDSS_r:rest'].value
    lum_tot = lum_disk + lum_sphere
    tot_ellip =  (lum_disk*ellip_disk + lum_sphere*ellip_spheroid)/(lum_tot)
```
So I think this should resolve your problems with that quantity.  (FYI Total size is a similarly weighted sum of the disk and bulge sizes). Please let me know if you think there is still a problem.

**Reply by yuemh** [reply]
Hi Eve,

Thank you again for all your help!

After discussing with other people in my group, I now have the feeling that the two definitions of ellipticity, e'=(1-q^2)/(1+q^2) and e=sqrt(e'), might get mixed up somewhere. I am pretty sure that the distribution of e=sqrt(e') in CosmoDC2 follows the Johnson SB distribution in Korytov+2019. But the Johnson SB functions themselves look like the distribution of e' in Joachimi+2013 (their Figure 7). If we instead make the distribution of e'=(1-q^2)/(1+q^2) of mock galaxies follow the Johnson SB distributions, the suspicious peak at q~1 disappears. So I think it might be worthy to check how the ellipticities are generated in CosmoDC2. Anyway, a sample with almost all galaxies being round is weird.

I really learned a lot from your suggestions. I appreciate that very much!

Minghao

**Reply by evevkovacs** [reply]
Hi Minghao, We are looking into this now.

**Reply by evevkovacs** [reply]
Hi Mingahao,
We have confirmed that there is a bug in the ellipticity distributions in the cosmoDC2 catalog.  In the catalog production code, the Johnson B distributions were correctly generated for the disk and bulge ellipticities. However, there was an error in computing the axis ratio from the ellipticity.  Specifically, the bug was caused by the following error in the code to compute the axis ratios from the generated distribution:
q_{bad} = sqrt((1 - e^2)/(1 + e^2))
The correct relationship should have been
q = sqrt((1 - e)/(1 + e))
We are in the process of deciding how best to address this issue and how to correct the catalog. I have new code to change the GCR reader to return the corrected values but we need to decide how best to deploy it.
 In the meantime,  you can generate corrected values by backing out the wrong definition to infer the Johnson B values and then re-deriving the correct values. Don't forget that the ellipticity values returned by the GCR use the shear ellipticity definition of (1-q)/(1+q).  Alternatively, there are native quantities in the catalog called 'morphology/diskAxisRatio' and 'morphology/sheroidAxisRatio'.  These are incorrect values, but you can use them to infer the original Johnson B values for the disk and bulge components and then recompute the correct axis ratios etc. 

Thank you so much for pursuing this and bringing the problem to our attention. We have acknowledged you in the validation paper that will be on the archive shortly.  Please let me know if you have any questions and thank you again for pointing out the problem.

**Reply by yuemh** [reply]
Hi Eve,

Thank you so much for your help! This info is super helpful to my project. Looking forward to the upcoming validation paper!

Minghao

**Reply by evevkovacs** [reply]
FYI, here's a plot showing the cross check that I made on the ellipticity calculations. The code for the modified reader is not available yet. Let me know if you have any issues reproducing this. 
![image](https://user-images.githubusercontent.com/7797837/135685457-329859a0-1ae6-4016-a822-a5a7ac5df995.png)
The distributions in this plot are offset by .01 so you can see that they match. I'm happy to provide more details if you want.

**Reply by evevkovacs** [reply]
And here is a comparison of the corrected and uncorrected distributions:
![image](https://user-images.githubusercontent.com/7797837/135688818-883786d5-5862-4278-a4ba-542f633b6bea.png)
